Allow specification of resource requirements in tools.

jmchilton opened the pull request:
Backend doesn't yet support them (except for providing a list of `ResourceRequirement` on the tool instance, so dynamic job rules could use them), but allow tool authors to specify what the runtime should support. Based on CWL resource requirements (http://www.commonwl.org/v1.2/CommandLineTool.html#ResourceRequirement).

xref https://github.com/galaxyproject/galaxy/issues/5369

jmchilton committed:
Allow specification of resource requirements in tools.

Backend doesn't yet support them. Based on CWL resource requirements (http://www.commonwl.org/v1.0/CommandLineTool.html#ResourceRequirement).

bernt-matthias commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -322,6 +322,36 @@ serve as complete descriptions of the runtime of a tool.
       <xs:element name="requirement" type="Requirement" minOccurs="0" maxOccurs="unbounded"/>
       <xs:element name="container" type="Container" minOccurs="0" maxOccurs="unbounded"/>
     </xs:sequence>
+    <xs:attribute name="cores_min" type="xs:long" use="required">
```
`use="required"` is making it mandatory to specify a value if the attribute is set, or?

mvdbeek replied:
```suggestion
    <xs:attribute name="cores_min" type="xs:long" use="optional">
```

mvdbeek commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -322,6 +322,36 @@ serve as complete descriptions of the runtime of a tool.
+Maximum reserved number of CPU cores if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="cores_max" type="xs:long" use="required">
```
```suggestion
    <xs:attribute name="cores_max" type="xs:long" use="optional">
```

mvdbeek commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -322,6 +322,36 @@ serve as complete descriptions of the runtime of a tool.
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved number of CPU cores if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    <xs:attribute name="ram_min" type="xs:long" use="required">
```
```suggestion
    <xs:attribute name="ram_min" type="xs:long" use="optional">
```

mvdbeek commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -322,6 +322,36 @@ serve as complete descriptions of the runtime of a tool.
+Minimum reserved RAM in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="ram_max" type="xs:long" use="required">
```
```suggestion
    <xs:attribute name="ram_max" type="xs:long" use="optional">
```

mvdbeek commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -322,6 +322,36 @@ serve as complete descriptions of the runtime of a tool.
+Maximum reserved RAM in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="tmpdir_min" type="xs:long" use="required">
```
```suggestion
    <xs:attribute name="tmpdir_min" type="xs:long" use="optional">
```

mvdbeek commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -322,6 +322,36 @@ serve as complete descriptions of the runtime of a tool.
+Minimum reserved filesystem based storage for the designated temporary directory, in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="tmpdir_max" type="xs:long" use="required">
```
```suggestion
    <xs:attribute name="tmpdir_max" type="xs:long" use="optional">
```

mvdbeek commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -614,6 +614,36 @@ serve as complete descriptions of the runtime of a tool.
       <xs:element name="requirement" type="Requirement" minOccurs="0" maxOccurs="unbounded"/>
       <xs:element name="container" type="Container" minOccurs="0" maxOccurs="unbounded"/>
     </xs:sequence>
+    <xs:attribute name="cores_min" type="xs:long" use="optional">
```
Looks like xs:long isn't a valid type, but the [CWL 1.2 spec](https://www.commonwl.org/v1.2/CommandLineTool.html#ResourceRequirement) allows this to be floats, so maybe we should go with
```suggestion
    <xs:attribute name="cores_min" type="float" use="optional">
```
?

dannon committed:
Update lib/galaxy/tool_util/xsd/galaxy.xsd

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

mvdbeek committed:
Fix xml attribute type

mvdbeek committed:
Trim trailing whitespace from galaxy.xsd

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
+]]></xs:documentation>
+    <xs:attribute name="ram_min" type="xs:float" use="optional">
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved RAM in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
```
```suggestion
Minimum reserved RAM in mebibytes (2**20 bytes), if runtime allows it (not yet implemented in Galaxy).
```

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
+    <xs:attribute name="cores_max" type="xs:float" use="optional">
+    </xs:attribute>
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved number of CPU cores if runtime allows it (unimplemented in Galaxy currently).
```
```suggestion
Maximum reserved number of CPU cores, if runtime allows it (not yet implemented in Galaxy).
```

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
     </xs:sequence>
+    <xs:attribute name="cores_min" type="xs:float" use="optional">
+      <xs:documentation xml:lang="en"><![CDATA[
+Maximum reserved number of CPU cores if runtime allows it (unimplemented in Galaxy currently).
```
```suggestion
Minimum reserved number of CPU cores, if runtime allows it (not yet implemented in Galaxy).
```

mvdbeek replied:
I'm redoing this, I don't think we'd want to write javascript expressions within attributes.
I'm thinking of this:
```xml
    <requirements>
        <resource type="cores_min">1.1</resource>
        <resource type="ram_max"><![CDATA[runtime.cores * 2]]></resource>
    </requirements>
```

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
+    </xs:attribute>
+    <xs:attribute name="ram_max" type="xs:float" use="optional">
+      <xs:documentation xml:lang="en"><![CDATA[
+Maximum reserved RAM in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
```
```suggestion
Maximum reserved RAM in mebibytes (2**20 bytes), if runtime allows it (not yet implemented in Galaxy).
```

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
+    </xs:attribute>
+    <xs:attribute name="tmpdir_min" type="xs:float" use="optional">
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved filesystem based storage for the designated temporary directory, in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
```
```suggestion
Minimum reserved filesystem-based storage for the designated temporary directory in mebibytes (2**20 bytes), if runtime allows it (not yet implemented in Galaxy).
```

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
+    </xs:attribute>
+    <xs:attribute name="tmpdir_max" type="xs:float" use="optional">
+      <xs:documentation xml:lang="en"><![CDATA[
+Maximum reserved filesystem based storage for the designated temporary directory, in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
```
```suggestion
Maximum reserved filesystem-based storage for the designated temporary directory in mebibytes (2**20 bytes), if runtime allows it (not yet implemented in Galaxy).
```

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="cores_max" type="xs:float" use="optional">
+    </xs:attribute>
```
```suggestion
```

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
+    </xs:attribute>
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved number of CPU cores if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
```
```suggestion
]]></xs:documentation>
    </xs:attribute>
```

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
+Maximum reserved filesystem based storage for the designated temporary directory, in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    </xs:attribute>
+  </xs:attribute>
```
```suggestion
```

mvdbeek committed:
Add `<resource/>` tag set to schema, add tool

I think this leaves us more room to write expressions without having to
escape everything.

mvdbeek committed:
Parse resource requirements

This makes resource requirements availbale on the tool instance,
so dyanmic job runners could start dispatching on these.

mvdbeek committed:
Use unimplemented instead of not yet implemented

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -6724,6 +6746,43 @@ and ``bibtex`` are the only supported options.</xs:documentation>
+      </xs:enumeration>
+      <xs:enumeration value="tmpdir_max">
+        <xs:annotation>
+          <xs:documentation xml:lang="en">Maximum reserved filesystem based storage for the designated temporary directory, in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).</xs:documentation>
```
```suggestion
          <xs:documentation xml:lang="en">Maximum reserved filesystem based storage for the designated temporary directory in mebibytes (2**20 bytes), if runtime allows it (unimplemented in Galaxy currently).</xs:documentation>
```

jdavcs reviewed (approved):
LGTM, except edits by @nsoranzo cannot be automatically applied - they should be included before we merge this.

mvdbeek committed:
Use 'not yet implemented in Galaxy'

nsoranzo commented on test/functional/tools/resource_requirements.xml:
```diff
@@ -0,0 +1,19 @@
+<tool id="resource_requirements" name="resource_requirements" version="1.0.0" profile="22.05">
+    <requirements>
+        <resource type="cores_min">1.1</resource>
```
Here you're using `resource` while in other places the element is `resource_requirement`. I think `resource` may be better (shorter and already inside `requirements`).

mvdbeek replied:
For Xml tools yes, for CWL and YAML I don't think they should go under requirements.

mvdbeek replied:
Hmm, okay, no, we should put them under requirement

mvdbeek replied:
for cwl ... I guess for yaml tools the idea was to have additional requirement types at the root level, like we do for containers. not sure that was a good idea though

nsoranzo commented on lib/galaxy/tool_util/deps/requirements.py:
```diff
@@ -1,4 +1,11 @@
 import copy
+from enum import Enum
```
```suggestion
```

mvdbeek replied:
What's the idea here ? Don't we need this for the ResourceType ?

nsoranzo commented on lib/galaxy/tool_util/deps/requirements.py:
```diff
@@ -1,4 +1,11 @@
+    Dict,
+    Optional,
+    Union,
+)
```
```suggestion
)

from typing_extensions import (
    get_args,
    Literal,
)

```

mvdbeek replied:
Not using Literal or get_args ... I suppose there's a suggestion missing here ?

nsoranzo commented on lib/galaxy/tool_util/deps/requirements.py:
```diff
@@ -252,10 +306,20 @@ def parse_requirements_from_xml(xml_root):
 
+def resource_from_element(resource_elem):
+    value_or_expression = xml_text(resource_elem)
+    resource_type = resource_elem.get("type", DEFAULT_CONTAINER_TYPE)
```
```suggestion
    resource_type = resource_elem.get("type")
```

nsoranzo commented on lib/galaxy/tool_util/deps/requirements.py:
```diff
@@ -204,20 +211,67 @@ def __str__(self):
+class ResourceRequirement:
+    def __init__(self, value_or_expression: Union[int, float, str], resource_type: ResourceType):
+        self.value_or_expression = value_or_expression
+        self.resource_type = resource_type
```
```suggestion
        if not resource_type:
            raise ValueError("Missing resource requirement type")
        if resource_type not in VALID_RESOURCE_TYPES:
            raise ValueError(f"Invalid resource requirement type '{resource_type}'")
        self.resource_type = resource_type
```

mvdbeek replied:
Again, `VALID_RESOURCE_TYPES` missing ?

nsoranzo replied:
Mmmh, one suggestion got lost somehow, re-added now.

nsoranzo commented on lib/galaxy/tool_util/deps/requirements.py:
```diff
@@ -204,20 +211,67 @@ def __str__(self):
+    ram_min = "ram_min"
+    ram_max = "ram_max"
+    tmpdir_min = "tmpdir_min"
+    tmpdir_max = "tmpdir_max"
```
```suggestion
ResourceType = Literal[
    "cores_min",
    "cores_max",
    "ram_min",
    "ram_max",
    "tmpdir_min",
    "tmpdir_max",
]
VALID_RESOURCE_TYPES = get_args(ResourceType)
```

mvdbeek committed:
Replace ResourceType Enum with Literal

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek committed:
Use resource consistently

mvdbeek commented on lib/galaxy/tool_util/deps/requirements.py:
```diff
@@ -211,18 +216,24 @@ def __str__(self):
-    ram_max = "ram_max"
-    tmpdir_min = "tmpdir_min"
-    tmpdir_max = "tmpdir_max"
+ResourceType = Literal[
```
How do I use this with non-literal values ? Why is this better than a Enum ?

nsoranzo replied:
It's currently only used for static checking, what do you have in mind?

mvdbeek replied:
https://github.com/galaxyproject/galaxy/pull/5378/commits/6972f0ffad996257065bc578992cde25f38126b9

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -680,6 +691,17 @@ resolver.
+      <xs:extension base="xs:string">
+        <xs:attribute name="type" type="ResourceType" use="required">
+          <xs:annotation>
+            <xs:documentation xml:lang="en">This value describes the resource available to the tool at runtime. Currently not implemented in Galaxy.</xs:documentation>
```
```suggestion
            <xs:documentation xml:lang="en">This value describes the type of resource required by the tool at runtime. Not yet implemented in Galaxy.</xs:documentation>
```

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -598,11 +598,22 @@ practice.
     </xs:attribute>
   </xs:complexType>
 
+  <xs:complexType name="ResourceRequirement">
```
Is this still used?

mvdbeek committed:
Improve ResourceType docs

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -598,11 +598,22 @@ practice.
+      <xs:element name="tmpdir_max" type="Resource" minOccurs="0" maxOccurs="1" />
+    </xs:sequence>
+  </xs:complexType>
+
```
```suggestion
```

mvdbeek committed:
Drop ResourceRequirement complex type

Not using this

nsoranzo commented on lib/galaxy/tool_util/deps/requirements.py:
```diff
@@ -204,20 +217,97 @@ def __str__(self):
+                self._runtime_required = False
+            except ValueError:
+                self._runtime_required = True
+        return self._runtime_required
```
Can we simplify here?

```suggestion
        try:
            float(self.value_or_expression)
            self.runtime_required = False
        except ValueError:
            self.runtime_required = True
```

nsoranzo commented on lib/galaxy/tool_util/parser/interface.py:
```diff
@@ -182,7 +182,7 @@ def parse_required_files(self) -> Optional["RequiredFiles"]:
     @abstractmethod
     def parse_requirements_and_containers(self):
-        """Return pair of ToolRequirement and ContainerDescription lists."""
+        """Return triple of ToolRequirement, ContainerDescription and ResourceRequirements lists."""
```
```suggestion
        """Return triple of ToolRequirement, ContainerDescription and ResourceRequirement lists."""
```

nsoranzo commented on lib/galaxy/util/xml_macros.py:
```diff
@@ -5,10 +5,12 @@
+try:
+    from lxml.etree import _Element
+except ImportError:
+    pass
```
What error is aabac58533fdb5aa2da7dd679a562aad370c9bb3 fixing exactly? The galaxy-util package doesn't require lxml currently, so mypy should fail??

mvdbeek replied:
Hmm, mypy was complaining that you can't use a variable for type annotations ... but I reverted the commit and mypy doesn't fail, so ... 🤷

mvdbeek replied:
Alright, seeing that again locally:
```
lib/galaxy/util/xml_macros.py:29: error: Variable "galaxy.util.Element" is not valid as a type
[valid-type]
        macros: Dict[str, List[Element]] = {}
                               ^
lib/galaxy/util/xml_macros.py:29: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
lib/galaxy/util/xml_macros.py:36: error: Element? has no attribute "get"  [attr-defined]
            tokens[m.get("name")] = m.text or ""
                   ^
lib/galaxy/util/xml_macros.py:36: error: Element? has no attribute "text"  [attr-defined]
            tokens[m.get("name")] = m.text or ""
                                    ^
lib/galaxy/util/xml_macros.py:42: error: Element? has no attribute "get"  [attr-defined]
            macro_dict[m.get("name")] = XmlMacroDef(m)
                       ^
Found 4 errors in 1 file (checked 1460 source files)
```

nsoranzo replied:
Probably some different package versions w.r.t. lib/galaxy/dependencies/pinned-lint-requirements.txt ?

mvdbeek replied:
absolutely possible, but it also makes sense, doesn't it ? What do you think about https://github.com/galaxyproject/galaxy/pull/14197 ?

mvdbeek committed:
Infer `runtime_required` just once

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek committed:
Another typo

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>
